platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerDaemonProcessFailureIntegrationTest.groovy

```diff
@@ -0,0 +1,288 @@
+            @WorkerPid
+            public class Bar {
+                public static void main(String[] args) {
+                    System.out.println("Hello world!");
+                }
+            }
+        """
+        def handler = blockingHttpServer.expectAndBlock("between")
+
+        when:
+        def gradle = executer.withTasks("compileJava").start()
+
+        then:
+        handler.waitForAllPendingCalls()
+
+        and:
+        pidFile('other').exists()
+        def pid1 = pidFile('other').text.strip() as long
+        new ProcessFixture(pid1).kill(false)
+
+        when:
+        handler.releaseAll()
+        result = gradle.waitForFinish()
+
+        then:
+        pidFile().exists()
+        def pid2 = pidFile().text.strip() as long
+        pid2 != pid1
+
+        and:
+        outputContainsKilledWorkerWarning()
+    }
+
+    void "daemon is gracefully removed if it is killed while idle before clients are stopped"() {
+        given:
+        buildFile << """
+            plugins {
+                id 'java'
+            }
+
+            ${dependsOnPidCapturingAnnotationProcessor}
+
+            tasks.withType(JavaCompile).configureEach {
+                options.fork = true
+                doLast {
+                    ${blockingHttpServer.callFromTaskAction('after')}
+                    services.get(org.gradle.workers.internal.WorkerDaemonFactory.class)
+                        .clientsManager.selectIdleClientsToStop { clients -> clients }
+                }
+            }
+        """
+        file('src/main/java/Foo.java') << """
+            @WorkerPid
+            public class Foo {
+                public static void main(String[] args) {
+                    System.out.println("Hello world!");
+                }
+            }
+        """
+        def handler = blockingHttpServer.expectAndBlock("after")
+
+        when:
+        def gradle = executer.withTasks("compileJava").start()
+
+        then:
+        handler.waitForAllPendingCalls()
+
+        and:
+        pidFile().exists()
+        def pid1 = pidFile().text.strip() as long
+        new ProcessFixture(pid1).kill(false)
+
+        when:
+        handler.releaseAll()
+        result = gradle.waitForFinish()
+
+        then:
+        outputContainsKilledWorkerWarning()
+    }
+
+    void outputContainsKilledWorkerWarning() {
+        if (OperatingSystem.current().windows) {
+            outputContains(" exited unexpectedly with exit code 1.")
+        } else {
+            outputContains(" exited unexpectedly after being killed with signal 9.  This is likely because an external process has killed the worker.")
+        }
+    }
+
+    TestFile pidFile(String rootPath = null) {
+        def root = rootPath ? file(rootPath) : testDirectory
+        return root.file('build/generated/sources/annotationProcessor/java/main/resources/pid.txt')
+    }
+
+    String getDependsOnPidCapturingAnnotationProcessor() {
```

Comment:
💅 This seems like a lot for a String-returning method to do, and I was a little confused reading through the tests how this would do something with an annotation processor.  Maybe most of it should be moved to a `setup` and just the dependencies bit remain here.

Reply:
Yeah, this method is probably a bit much now that you point it out.  I'll look at cleaning this up a little.